test/unit/org/apache/cassandra/cql3/validation/operations/InsertInvalidateSizedRecordsTest.java

```diff
@@ -0,0 +1,195 @@
+
+        // null / empty checks
+        Assertions.assertThatThrownBy(() -> executeNet("INSERT INTO %s (a, b) VALUES (?, ?)", MEDIUM_BLOB, null))
+                  .hasRootCauseInstanceOf(InvalidQueryException.class)
+                  .hasRootCauseMessage("Invalid null value in condition for column b");
+
+        // org.apache.cassandra.db.MultiCBuilder.OneClusteringBuilder.addElementToAll defines "null" differently than most of the code;
+        // most of the code defines null as:
+        //   value == null || accessor.isEmpty(value)
+        // but the code defines null as
+        //   value == null
+        // In CASSANDRA-18504 a new isNull method was added to the type, as blob and text both "should" allow empty, but this scattered null logic doesn't allow...
+        // For backwards compatability reasons, need to keep empty support
+        executeNet("INSERT INTO %s (a, b) VALUES (?, ?)", MEDIUM_BLOB, EMPTY_BYTE_BUFFER);
+    }
+
+    @Test
+    public void compositeValueClustering()
+    {
+        createTable(KEYSPACE, "CREATE TABLE %s (a blob, b blob, c blob, PRIMARY KEY (a, b, c))");
+        // sum of columns is too large
+        Assertions.assertThatThrownBy(() -> executeNet("INSERT INTO %s (a, b, c) VALUES (?, ?, ?)", MEDIUM_BLOB, MEDIUM_BLOB, MEDIUM_BLOB))
+                  .hasRootCauseInstanceOf(InvalidQueryException.class)
+                  .hasRootCauseMessage("Key length of " + (MEDIUM_BLOB.remaining() * 2) + " is longer than maximum of 65535");
+
+        // single column is too large
+        // the logic prints the total clustering size and not the single column's size that was too large
+        Assertions.assertThatThrownBy(() -> executeNet("INSERT INTO %s (a, b, c) VALUES (?, ?, ?)", MEDIUM_BLOB, MEDIUM_BLOB, LARGE_BLOB))
+                  .hasRootCauseInstanceOf(InvalidQueryException.class)
+                  .hasRootCauseMessage("Key length of " + (MEDIUM_BLOB.remaining() + LARGE_BLOB.remaining()) + " is longer than maximum of 65535");
+    }
+
+    @Test
+    public void singleValueIndex()
+    {
+        createTable(KEYSPACE, "CREATE TABLE %s (a blob, b blob, PRIMARY KEY (a))");
+        String table = KEYSPACE + "." + currentTable();
+        execute("CREATE INDEX single_value_index ON %s (b)");
+        Assertions.assertThatThrownBy(() -> executeNet("INSERT INTO %s (a, b) VALUES (?, ?)", MEDIUM_BLOB, LARGE_BLOB))
+                  .hasRootCauseInstanceOf(InvalidQueryException.class)
+                  .hasRootCauseMessage("Cannot index value of size " + LARGE_BLOB.remaining() + " for index single_value_index on " + table + "(b) (maximum allowed size=65535)");
+    }
+
+    /**
+     * Partial port of cqlsh_tests.test_cqlsh_copy.TestCqlshCopy#test_reading_empty_strings_for_different_types
+     */
+    @Test
+    public void test_reading_empty_strings_for_different_types() throws IOException
```

Comment:
I think we can use a camel cased method name, the JavaDoc still makes it easy to find the corresponding Python dtest.